packages/@aws-cdk/integ-tests/lib/assertions/waiter-state-machine.ts

```diff
@@ -0,0 +1,146 @@
+import { CfnResource, Duration, Stack } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { AssertionsProvider } from './providers';
+
+/**
+ * Options for creating a WaiterStateMachine
+ */
+export interface WaiterStateMachineOptions {
+  /**
+   * The total time that the state machine will wait
+   * for a successful response
+   *
+   * @default Duration.minutes(30)
+   */
+  readonly totalTimeout?: Duration;
+
+  /**
+   * The interval to wait between attempts.
+   *
+   * @default Duration.seconds(5)
+   */
+  readonly interval?: Duration;
+
+  /**
+   * Backoff between attempts.
+   *
+   * @default 1
+   */
+  readonly backoffRate?: number;
+}
+
+/**
+ * Props for creating a WaiterStateMachine
+ */
+export interface WaiterStateMachineProps extends WaiterStateMachineOptions {}
+
+/**
+ * A very simple StateMachine construct highly customized to the provider framework.
+ * This is so that this package does not need to depend on aws-stepfunctions module.
+ *
+ * The state machine continuously calls the isCompleteHandler, until it succeeds or times out.
+ * The handler is called `maxAttempts` times with an `interval` duration and a `backoffRate` rate.
```

Comment:
Is it worth having it be incremented?
`backoffRate` going from 1, to 2, to 3, to 5 etc?

If it does not change, why have it at all in the current context?

Reply:
I'll add some more documentation around the backoffRate. It corresponds to the BackoffRate of the StepFunctions Retry state https://docs.aws.amazon.com/step-functions/latest/dg/concepts-error-handling.html#error-handling-retrying-after-an-error.

Since we are using the Retry state as a waiter, most of the time we don't actually want any backoff, we just want to retry every 5 seconds (or whatever the interval is). If for whatever reason the user _does_ want to add a backoff rate they can.

Reply:
Added some docs as well as updated the logic to incorporate the backoff rate in the maxRetries calculation.